core/src/commonMain/kotlin/com/sourcepoint/mobile_core/Coordinator.kt

```diff
@@ -449,151 +394,138 @@ class Coordinator(
         }
     }
 
-    suspend fun pvData(pubData: JsonObject?, messages: List<MessageToDisplay>) {
-        coroutineScope {
-            if (campaigns.gdpr != null && state.gdprMetaData != null) {
-                launch {
-                    val sampled = sampleAndPvData(
-                        state.gdprMetaData!!,
-                        gdprPvDataBody(
-                            consent = state.gdpr,
-                            pubData = pubData,
-                            messageMetaData = messages.first { it.type == SPCampaignType.Gdpr }.metaData
-                        ))
-                    state.gdprMetaData = state.gdprMetaData!!.copy(wasSampled = sampled)
-                }
+    private suspend fun pvData(pubData: JsonObject?, messages: List<MessageToDisplay>) = coroutineScope {
+        val gdprPvData = campaigns.gdpr?.let {
+            launch {
+                gdprPvData(pubData, messageMetaData = messages.first { it.type == SPCampaignType.Gdpr }.metaData)
             }
-            if (campaigns.ccpa != null && state.ccpaMetaData != null) {
-                launch {
-                    val sampled = sampleAndPvData(
-                        state.ccpaMetaData!!,
-                        ccpaPvDataBody(
-                            consent = state.ccpa,
-                            pubData = pubData,
-                            messageMetaData = messages.first { it.type == SPCampaignType.Ccpa }.metaData
-                        ))
-                    state.ccpaMetaData = state.ccpaMetaData!!.copy(wasSampled = sampled)
-                }
+        }
+        val ccpaPvData = campaigns.ccpa?.let {
+            launch {
+                ccpaPvData(pubData, messageMetaData = messages.first { it.type == SPCampaignType.Ccpa }.metaData)
             }
-            if (campaigns.usnat != null && state.usNatMetaData != null) {
-                launch {
-                    val sampled = sampleAndPvData(
-                        state.usNatMetaData!!,
-                        usnatPvDataBody(
-                            consent = state.usNat,
-                            pubData = pubData,
-                            messageMetaData = messages.first { it.type == SPCampaignType.UsNat }.metaData
-                        ))
-                    state.usNatMetaData = state.usNatMetaData!!.copy(wasSampled = sampled)
-                }
+        }
+        val usNatPvData = campaigns.usnat?.let {
+            launch {
+                usnatPvData(pubData, messageMetaData = messages.first { it.type == SPCampaignType.Ccpa }.metaData)
```

Comment:
My reaction: https://www.youtube.com/shorts/OmCvkZ-mc-8
I believe `SPCampaignType.Ccpa` should be `SPCampaignType.UsNat` 
I'm glad I've re-read this class twice.